tests/bluetooth/bsim_bt/bsim_test_eatt/src/main_collision.c

```diff
@@ -0,0 +1,224 @@
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Disconnected: %s (reason 0x%02x)\n", addr, reason);
+
+	bt_conn_unref(default_conn);
+	default_conn = NULL;
+	is_connected = false;
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+
+static void test_peripheral_main(void)
+{
+	int err;
+	size_t num_enhanced;
+
+	err = bt_enable(NULL);
+	if (err) {
+		FAIL("Can't enable Bluetooth (err %d)\n", err);
+	}
+
+	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
+	if (err) {
+		FAIL("Advertising failed to start (err %d)\n", err);
+	}
+
+	while (!is_connected) {
+		k_sleep(K_MSEC(100));
+	}
+
+	err = bt_eatt_connect(default_conn, CONFIG_BT_EATT_MAX);
+	if (err) {
+		FAIL("Sending credit based connection request failed (err %d)\n", err);
+	}
+
+	/* TODO: Use a flag set by a callback from EATT connection instead of sleeping */
+	/* Wait for EATT channels to be connected */
+	k_sleep(K_MSEC(10000));
+	num_enhanced = bt_eatt_count(default_conn);
+	printk("%d enhanced bearers connected\n", num_enhanced);
+	if (num_enhanced != CONFIG_BT_EATT_MAX) {
+		FAIL("Expected %d enhanced bearers, got %d\n", CONFIG_BT_EATT_MAX, num_enhanced);
+	}
+	/* Disconnect */
+	err = bt_conn_disconnect(default_conn, BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+	if (err) {
+		FAIL("Disconnection failed (err %d)\n", err);
+	}
+
+	while (is_connected) {
+		k_sleep(K_MSEC(100));
+	}
+
+	PASS("EATT Peripheral tests Passed\n");
+}
+
+static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
+			 struct net_buf_simple *ad)
+{
+	int err;
+
+	err = bt_le_scan_stop();
+	if (err) {
+		FAIL("Stop LE scan failed (err %d)\n", err);
+	}
+
+	err = bt_conn_le_create(addr, BT_CONN_LE_CREATE_CONN, BT_LE_CONN_PARAM_DEFAULT,
+				&default_conn);
+	if (err) {
+		FAIL("Create conn failed (err %d)\n", err);
+	}
+
+	printk("Device connected\n");
+}
+
+static void test_central_main(void)
+{
+	int err;
+	size_t num_enhanced;
+
+	err = bt_enable(NULL);
+	if (err) {
+		FAIL("Can't enable Bluetooth (err %d)\n", err);
+	}
+
+	err = bt_le_scan_start(BT_LE_SCAN_ACTIVE, device_found);
+	if (err) {
+		FAIL("Scanning failed to start (err %d)\n", err);
+	}
+
+	while (!is_connected) {
+		k_sleep(K_MSEC(100));
+	}
+
+	err = bt_eatt_connect(default_conn, CONFIG_BT_EATT_MAX);
+	if (err) {
+		FAIL("Sending credit based connection request failed (err %d)\n", err);
+	}
+
+	/* TODO: Use a flag set by a callback from EATT connection instead of sleeping */
+	/* Wait for EATT channels to be connected */
+	k_sleep(K_MSEC(10000));
+	num_enhanced = bt_eatt_count(default_conn);
+	printk("%d enhanced bearers connected\n", num_enhanced);
+	if (num_enhanced != CONFIG_BT_EATT_MAX) {
+		FAIL("Expected %d enhanced bearers, got %d\n", CONFIG_BT_EATT_MAX, num_enhanced);
+	}
+
+	/* Wait for disconnect */
+	while (is_connected) {
+		k_sleep(K_MSEC(100));
+	}
+
+	PASS("EATT Central tests Passed\n");
```

Comment:
I assume that you've run this and seen the debug message that indicate a collision, right? Because I notice that this test doesn't actually wait for a collision before passing.

Reply:
A collision will happen because both central and peripheral sends the ecred connection request right after connection. We see this in the log because both fail to allocate ATT channels on receiving the request. Without the fix in this PR, both respond with "No resources" and do not retry the request. Thus zero EATT channels are created and the `num_enhanced != CONFIG_BT_EATT_MAX` check will be true and fail the test.